### PR TITLE
feat(igla-race): Path A · Phase B — flip TRIOS_USE_REAL_BPB, prove Δbpb ≥ 1e-6 (refs #446 #509 #521)

### LINKED ISSUE #509: 🚨 P0: Wave-3 anomaly — per-seed BPB identical across formats (fp32-fallback confirmed)
# 🚨 P0: Wave-3 Anomaly — Per-seed BPB Identical Across All Formats (fp32-fallback confirmed)

**Severity:** P0 — Invalidates all format-specific BPB claims
**Related:** #109 (corpus contamination), #507 (champion-loop deadlock), #446 (EPIC)
**R5 Classification:** tracking-anomalous

---

## What Happened

Wave-3 canary (378 experiments, 63 formats × 6 seeds) revealed that **BPB is determined solely by seed**, not by format:

| Seed | BPB @ step 3000 | # formats sharing this BPB |
|---:|---:|---:|
| 1597 | 2.942101001739502 | ≥52 |
| 10946 | 2.945727825164795 | ≥44 |
| 17711 | 2.963225364685059 | ≥46 |
| 6765 | 2.964042901992798 | ≥47 |
| 4181 | 2.998884916305542 | ≥49 |
| 2584 | 3.001356363296509 | ≥48 |

**6 distinct (seed, BPB) pairs across 286 completed canaries. Zero format-dependent variation.**

## Root Cause (CONFIRMED via code audit)

**The `format` parameter in `cpu_train.rs` is a LABEL ONLY — it is used for the output filename, NOT for any computation.**

Evidence from source code audit:

```
src/bin/cpu_train.rs:
  Line N: let format_type = std::env::var("TRIOS_FORMAT_TYPE").ok();
  Line N: // For Phase E.GF v2, use format type in result filename
  Line N: let format_suffix = format_type.as_ref().unwrap_or(&default_format);
  Line N: let rpath = format!("results/{}_seed{}.json", format_suffix, seed);
```

**ZERO calls to:** `quantize`, `gf16_encode`, `gf16_decode`, `phi_bias`, `PHI_BIAS`, `from_f32`, `to_f32`

**ALL computations in the training loop use `f32`:** `forward(&self, tokens: &[usize]) -> Vec<Vec<f32>>`, `step(&mut self, params: &mut [f32], grads: &[f32])`, `loss_and_grad(&self, tokens: &[usize]) -> (f32, ...)`

## Impact Assessment

| Claim | R5 Status |
|---|---|
| "gf16 = bf16 = fp16 = 2.5655 BPB" | ❌ **INVALID** — all three run identical f32 code, of course they match |
| "φ-distance predicts training quality" | ❌ **UNPROVEN** — we never actually tested different formats |
| "GF12 100% success rate" | ❌ **MISLEADING** — "success" means f32 training completed, not GF12-specific |
| "Posit8/16/32/64 measured" | ❌ **INVALID** — measured f32 with a posit label |
| "NF4/LNS/MXFP measured" | ❌ **INVALID** — same f32 code |
| Champion BPB = 2.2393 | ⚠ **LEGITIMATE** — but it's f32 performance, not format-specific |
| All 241 cells in phd_format_algo_matrix | ❌ **INVALID** — format dimension is noise |

## What's Real vs What's Not

| Dimension | Real? | Notes |
|---|---|---|
| **Seed** | ✅ | Different seeds → different BPB (confirmed) |
| **Optimizer** (adamw vs muon) | ✅ | Different optimizers → different BPB (confirmed) |
| **Hidden size** | ✅ | Different h → different BPB (confirmed) |
| **Learning rate** | ✅ | Different LR → different BPB (confirmed) |
| **Format** | ❌ | Format is a no-op label in current trainer |

### P0 — Immediate (before any more experiments)

- [x] **Prune 4,320 PHD-FULL-GRID pending experiments** — they're meaningless without format-specific computation
- [ ] **R5 reclassify champion** — BPB=2.2393 is f32 performance, format claims are void
- [ ] **Update #446 matrix** — add `format_effective=FALSE` flag to all cells
- [ ] **Stop claiming format-specific results** in any paper, PR, or comment

### P1 — Fix the trainer (root cause)

Two approaches:

**A. Add format-specific quantization to training loop:**
```rust
// In cpu_train.rs forward():
// After each matmul, quantize weights to target format then dequantize back
// This makes the forward pass format-aware
let quantized = phi_numbers::gf16::from_f32(weight);
let dequantized = phi_numbers::gf16::to_f32(quantized);
// Use dequantized for subsequent computation
```

**B. Post-training quantization (PTQ) approach:**
- Train in f32 (current behavior)
- After training, quantize the model to target format
- Measure BPB with quantized model (inference-only)
- This is what NF4/QLoRA/GPTQ do — it's legitimate

**Recommendation:** Approach B is faster to implement and matches industry practice. Approach A is more rigorous for PhD but requires significant code changes.

### P2 — Validation experiment (Wave-4)

After implementing format-aware training:
- [ ] Run 2-format × 1-seed micro-canary with `RUST_LOG=debug`
- [ ] Verify kernel path logs show format-specific code
- [ ] Verify BPB diverges between formats for same seed/h/lr/opt
- [ ] Only then re-seed full grid

## Neon Updates

- `phd_format_algo_matrix` — 241 cells now have `bpb_diverges_per_format = FALSE`
- `phd_format_registry` — `is_measurable` flag is INVALID for format-specific claims (measures f32, not format)
- All format-specific BPB claims in #446 comments are retroactively void

---

**R5 honest assessment:** The format × algorithm matrix we built is a **seed × optimizer × h × LR** matrix with a format label that had no computational effect. The dimensional analysis (seed, optimizer, hidden, LR) is valid. The format dimension is null.

This is a **methodology error**, not a data fabrication issue. The data is real — it's just measuring the wrong thing for the format comparison we claimed.

φ² + φ⁻² = 3 · TRINITY · HONEST ABOVE ALL

### LINKED ISSUE #531: Phase B · flip TRIOS_USE_REAL_BPB=1 in CI · prove |Δbpb(f32, gf16)| ≥ 1e-6
## Phase B · flip TRIOS_USE_REAL_BPB=1 in CI · prove |Δbpb(f32, gf16)| ≥ 1e-6

Tracker for the Phase-B follow-up to #519 (which landed the bridge in default-bit-exact mode).

### Scope

- Fix `crates/trios-igla-race/src/real_forward.rs`: `cpu_train` parses `--seed=N --steps=N` CLI args, not env vars. Switch to `.arg(...)`.
- Honour `TRIOS_TRAINER_CWD` so `cpu_train` finds `data/tinyshakespeare.txt`.
- Add `.github/workflows/path-a-real-bpb.yml`: hermetic CI lane that
  - checks out `trios-trainer-igla` at pinned SHA `73aae9ea8d74a100c1d8204ea26fbe7b82795629`,
  - synthesises a 32 KB deterministic corpus,
  - `cargo build --release --bin cpu_train`,
  - runs L-FH4 with `TRIOS_USE_REAL_BPB=1`,
  - re-confirms `test_v2_default_matches_legacy` (legacy bit-exact still holds when env unset).

### Acceptance — pre-registered (OS-001-forward-hook-bridge-v1)

- ✅ `|Δbpb(f32, gf16)| ≥ 1e-6` for at least one of seeds {42, 43, 44} → bridge is real
- ❌ `|Δbpb| == 0` across all three seeds → bridge is still fake → reopen #509-FABRICATED root cause

### Closes / Refs

- Part of #446 (matrix observability — final Phase-B closure)
- Resolves #509 (#509-FABRICATED — when L-FH4 green)
- Built on #521 (Phase-A wire)

Anchor: φ² + φ⁻² = 3 · Zenodo DOI 10.5281/zenodo.19227877 · TRINITY · NEVER STOP

### PR DESCRIPTION
**Pull request by gHashTag**
Closes #531
Part of #446
Resolves #509
Built on #521 #519

## Phase B · flip `TRIOS_USE_REAL_BPB=1` in CI · prove `|Δbpb(f32, gf16)| ≥ 1e-6`

DRAFT — opened to get CI signal on the L-FH4 falsifier with the real-forward bridge enabled.

### What this PR does

| Lane | File | Purpose |
|------|------|---------|
| L-FH3 fix | `crates/trios-igla-race/src/real_forward.rs` | `cpu_train` parses `--seed=N --steps=N` CLI args (`arg_or` parser), **not** env vars. Replace `.env(TRIOS_SEED/TRIOS_STEPS)` with `.arg(format!("--seed={}", …))`. |
| L-FH3 fix | (same file) | Honour `TRIOS_TRAINER_CWD`; fall back to `<bin>/../../..` so `cpu_train` resolves `data/tinyshakespeare.txt` regardless of caller cwd. |
| L-FH4 CI | `.github/workflows/path-a-real-bpb.yml` | Hermetic CI lane: checks out `trios-trainer-igla` @ pinned SHA `73aae9ea` (BL-5 mitigation); synthesises 32 KB deterministic corpus (no network); `cargo build --release --bin cpu_train`; runs L-FH4 with `TRIOS_USE_REAL_BPB=1`; re-confirms `test_v2_default_matches_legacy` (default bit-exact still holds). |

### Pre-registered acceptance (OS-001-forward-hook-bridge-v1)

- ✅ `|Δbpb(f32, gf16)| ≥ 1e-6` for ≥1 of seeds {42, 43, 44} → **bridge is real, #509-FABRICATED dispelled**
- ❌ `|Δbpb| == 0` across all three seeds → bridge still fake → reopen with new lane

### Why this is final closure for #446 / #509

`simulate_bpb_v2` + `forward_real_bpb` already in `main` from #519 (verified: 6 + 3 hits via Contents API; search index lagging). With this PR's CI lane, the falsifier executes against the real `cpu_train` binary on a pinned trainer-igla SHA. If green: format-dependent metrics are now CI-enforced as real. If red: pre-registration's reject band fires honestly.

### Cron `f541bb68` victory condition

Will fire `TASK SOLVED` once: (#446 closed) ∧ (#509 closed) ∧ (`simulate_bpb_v2` ∈ main) ∧ (`forward_real_bpb` ∈ main). Last two conditions already true.

Anchor: φ² + φ⁻² = 3 · Zenodo DOI 10.5281/zenodo.19227877 · TRINITY · NEVER STOP

### CODE REVIEW

**Comment by gHashTag** [comment]
## L-FH4 GREEN · |Δbpb| = 1.4e-3 (R5-honest)

Pre-registration: [`docs/preregistration/forward_hook_bridge_v1.md`](https://github.com/gHashTag/trios/blob/main/docs/preregistration/forward_hook_bridge_v1.md)
One-shot: `OS-001-forward-hook-bridge-v1`
Anchor: φ² + φ⁻² = 3 · DOI 10.5281/zenodo.19227877

### Falsifier outcome (CI run 25477706628 · job 74754709372 · SHA `fac92b0`)

```
seed= 42  f32=7.005800  gf16=7.004500  Δ=0.001300
seed= 43  f32=7.004700  gf16=7.003500  Δ=0.001200
seed= 44  f32=7.006300  gf16=7.004900  Δ=0.001400
L-FH4 ACCEPTED: max |Δbpb| = 0.001400 ≥ 0.000001
```

- **Accept band**: `|Δbpb| ≥ 1e-6` for ≥1 seed in {42,43,44} — **met on all 3** (`Δ ∈ [1.2e-3, 1.4e-3]`).
- **Reject band**: `|Δbpb| == 0` across all 3 — **not entered**.
- Margin to threshold: `1400×`. Not a boundary case (BL-3 not invoked).

### What this proves (and what it does NOT)

✅ Proven: `simulate_bpb_v2(use_gf16=true)` ≠ `simulate_bpb_v2(use_gf16=false)` when `TRIOS_USE_REAL_BPB=1`. The bridge from `trios-igla-race` → `cpu_train` is **not** closed-form. `fake_quantize_weights` is on the hot path.
✅ Proven: legacy `simulate_bpb_v2` default-off path remains bit-exact (`test_v2_default_matches_legacy` GREEN — Confirm legacy bit-exact preservation step).
✅ Proven: `cpu_train` honors `TRIOS_FORMAT_TYPE` (smoke 50-step run shows distinct `train_loss` between f32 and gf16 already at step 50).

❌ NOT proven by this PR alone:
- Champion BPB ≤ 2.2393 reproduced under f32 vs gf16 split. The 32 KB synthetic corpus produces `Init BPB = 7.0` and stays around 7.005 in 200 steps — far from convergence (this is intentional, the falsifier needs only a measurable Δ, not convergence).
- A1 watchdog flips off `MERGED_BUT_NO_HOOK` — needs a separate signal/handshake (tracked elsewhere).
- A2 harness gets all 38 cells unfabricated — needs the matrix runner to consume `simulate_bpb_v2` end-to-end.

### Closes / supersedes

- Closes #531 (forward-hook bridge falsifier missing in CI).
- Resolves #509 acceptance criterion 1 (real-forward bridge wired, format actually changes BPB).
- Part of #446 — does **not** close it: matrix-coverage and #509 cell-fabrication remain.

### Sibling collaboration

- Branch was a 2-agent loop: agent (this work item) + Trinity Grandmaster sibling. Both arrived at the pipe-parser fix independently within 3 minutes (sibling pushed first as `44c7cbc`; agent reset to it). YAML-script extraction (`bf5d285`) is sibling's; rustc unpin (`7c4c39b`) is agent's. R10 atomic-commit invariant respected throughout.

R5-honest: this is one wall down on the #446 stack, not the whole house. Ready for review and merge into main; on merge, sibling's PR-#446 watcher (cron `f541bb68`) should flip the `Phase B NOT-ARMED` watchdog.
